### ISSUE TITLE
Implement unpublish / redraft for changeset specs

### ISSUE THREAD

**Issue by eseliger** [sourcegraph/sourcegraph#17541]
Currently, changing `published` from `true` to `false` does nothing, which is counter-intuitive and doesn't support our claim that we "always try to reconcile the state given in the spec". Also, it might be a feature you want to have, closing open changesets, while retaining them so you can open them again later (ie. you found a problem with your campaign and want to temporarily suspend reviews).

That would include three additional state transitions of changesets:
- true => false
- draft => false
- true => draft

The first two are kind of the same, basically run operation `close`, but not operation detach on those changesets.
The third case requires us to add a reconciler operation `redraft`, and to implement a method on the ChangesetSources `RedraftChangeset`, which would turn the changeset into draft mode again. For GitLab this would be adding the `WIP: ` prefix again, for GitHub calling an API to convert the PR back to draft. (As of now, I think this API doesn't exist, though... it's only available in their UI.)

**Comment by eseliger** [comment]
Yes, the first two we could implement. We could also implement "redraft" for GitLab, just not for GitHub

**Comment by malomarrec** [comment]
We synced on this:
- the first 2 items are in scope.
- true => draft will be another ticket, blocked by GitHub

**Comment by mrnugget** [comment]
Requirement for https://github.com/sourcegraph/sourcegraph/issues/18277

**Comment by bahrmichael** [comment]
@malomarrec @eseliger Can you confirm if the first 2 items were implemented? I just tried it on S2 and didn't see a PR on GitHub being closed when I set `published: false`. Not sure if my expectations are right.

**Comment by eseliger** [comment]
No I don't think this is implemented
